### PR TITLE
Floria: add state precompiled contract

### PR DESCRIPTION
**Pull request by simonlechner**
This PR adds support for the state precompiled contract which is used to handle epochs to Floria.

### CODE REVIEW

File: go/processor/floria/precompiled.go [hunk 1]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
+			panic("unknown EvmWriter method")
+		}
+
+		*constID = make([]byte, len(method.ID))
+		copy(*constID, method.ID)
+	}
+}
+
+var ErrExecutionReverted = fmt.Errorf("execution reverted")
+var ErrOutOfGas = fmt.Errorf("out of gas")
```
Comment (HerbertJordan) [inline go/processor/floria/precompiled.go:102]: nit: constError?

File: go/processor/floria/precompiled.go [hunk 2]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
+
+func handleStatePrecompiledContract(state tosca.WorldState, sender tosca.Address, address tosca.Address, input []byte, gas tosca.Gas) (tosca.CallResult, bool) {
```
Comment (HerbertJordan) [inline go/processor/floria/precompiled.go:104]: `address` => `receiver`?

File: go/processor/floria/precompiled.go [hunk 3]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
+	}
+	gas, err := statePrecompiledContract(state, sender, input, gas)
+	return tosca.CallResult{
+		Success: err == nil,
+		Output:  nil,
+		GasLeft: tosca.Gas(gas),
+	}, true
+}
+
+func statePrecompiledContract(state tosca.WorldState, sender tosca.Address, input []byte, gas tosca.Gas) (tosca.Gas, error) {
```
Comment (HerbertJordan) [inline go/processor/floria/precompiled.go:116]: nit: `runStatePrecompiledContract`?

Also the other functions in this file have a 'state' prefix to -- I guess -- artificially group them and avoid name collisions. Two ideas to group them otherwise would be
- define an empty "StateContract" struct and make them members of this type
- define a package "state_contract" and move the implementation into this package

Wdyt?

File: go/processor/floria/precompiled.go [hunk 4]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
+	gas -= CreateGas
+	if len(input) != 64 {
+		return 0, ErrExecutionReverted
+	}
+
+	accountTo := tosca.Address(input[12:32])
+	accountFrom := tosca.Address(input[32+12 : 32+32])
+
+	code := state.GetCode(accountFrom)
+	if code == nil {
```
Comment (HerbertJordan) [inline go/processor/floria/precompiled.go:174]: curious: is this check needed or would the rest of the code also work with a `nil` value for the code?

File: go/processor/floria/precompiled.go (hunk 2 above)
Comment (HerbertJordan) [inline go/processor/floria/precompiled.go:104]: Please add documentation to this function and a link to the original definition.

File: go/processor/floria/processor.go [hunk 5]
```diff
@@ -24,6 +24,12 @@ const (
 	TxAccessListAddressGas    = 2400
 	TxAccessListStorageKeyGas = 1900
 
+	CallValueTransferGas = tosca.Gas(9000)
+	CreateGas            = tosca.Gas(32000)
+	CreateDataGas        = tosca.Gas(200)
+	MemoryGas            = tosca.Gas(3)
+	SstoreSetGasEIP2200  = tosca.Gas(20000)
```
Comment (HerbertJordan) [inline go/processor/floria/processor.go:31]: If these constants are only ever to be used by the state contract (and tests for it) I would suggest to move them to the corresponding file.

File: go/processor/floria/precompiled.go
Comment (HerbertJordan) [inline go/processor/floria/precompiled.go:1]: filename: `state_contract` or `state_precompiled_contract` to avoid confusion with other precompiled contracts?

File: go/processor/floria/precompiled_test.go [hunk 6]
```diff
@@ -80,3 +81,347 @@ func TestPrecompiled_AddressesAreHandledCorrectly(t *testing.T) {
...
+			}
+			if gas != 0 {
+				t.Errorf("unexpected gas, want %v, got %v", 0, gas)
+			}
+		})
+	}
+}
+
+func TestPrecompiled_StateCopyCode(t *testing.T) {
+	codeSize := tosca.Gas(42)
```
Comment (HerbertJordan) [inline go/processor/floria/precompiled_test.go:138]: I guess a `codeSize` should not be of type `tosca.Gas`, right? (also in other tests)

File: go/processor/floria/precompiled_test.go [hunk 7]
```diff
@@ -80,3 +81,347 @@ func TestPrecompiled_AddressesAreHandledCorrectly(t *testing.T) {
...
+			},
+		},
+	}
+	for testName, test := range tests {
+		t.Run(testName, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			state := tosca.NewMockWorldState(ctrl)
+			test.mockSetUp(state)
+
+			sender := DriverAddress
```
Comment (HerbertJordan) [inline go/processor/floria/precompiled_test.go:414]: Is there a test checking that the precompiled contract is not executed if the sender is != DriverAddress?

File: go/processor/floria/precompiled_test.go [hunk 8]
```diff
@@ -80,3 +81,347 @@ func TestPrecompiled_AddressesAreHandledCorrectly(t *testing.T) {
...
+	tests := map[string]struct {
+		recipient          tosca.Address
+		inputPrefix        []byte
+		input              []byte
+		isStatePrecompiled bool
+		mockSetUp          func(*tosca.MockWorldState)
+	}{
+		"nonPrecompiled": {
+			tosca.Address{0x20},
+			[]byte{0x01, 0x02, 0x03, 0x04},
```
Comment (HerbertJordan) [inline go/processor/floria/precompiled_test.go:352]: Is there a test that checks that an insufficient long prefix is handled correctly?

File: go/processor/floria/run_context.go [hunk 9]
```diff
@@ -74,7 +74,13 @@ func (r runContext) Call(kind tosca.CallKind, parameters tosca.CallParameters) (
...
 	}
 
-	output, isPrecompiled := handlePrecompiledContract(r.blockParameters.Revision, parameters.Input, recipient, parameters.Gas)
+	output, isStatePrecompiled := handleStatePrecompiledContract(
+		r, parameters.Sender, parameters.Recipient, parameters.Input, parameters.Gas)
+	if isStatePrecompiled {
+		return output, nil
+	}
+	output, isPrecompiled := handlePrecompiledContract(
+		r.blockParameters.Revision, parameters.Input, recipient, parameters.Gas)
```
Comment (HerbertJordan) [inline go/processor/floria/run_context.go:83]: A heads-up for when you are implementing warm/cold accounts: If I remember the opera modification in go-ethereum correctly, there is a "bug" in the sense that only real pre-compiled contracts are marked as warm before an execution. The state-precompiled contract is missed there. When you are getting there, please check this carefully.

File: go/processor/opera/processor.go [hunk 10]
```diff
@@ -479,7 +479,8 @@ func init() {
 		}
 
 		*constID = make([]byte, len(method.ID))
-		copy(*constID, method.ID)
+		id := method.ID
+		copy(*constID, id)
```
Comment (HerbertJordan) [inline go/processor/opera/processor.go:483]: nit: unrelated debug change?

**Review by HerbertJordan — APPROVED** [review]
Looks good, except the crazy stuff that is there because it is defined the same way in the Opera version of this contract (e.g. the Gas price computation for copying codes).

I added mostly style comments, please have a look. In general, however, I think it is a clear improvement over the original.

File: go/processor/floria/precompiled.go [hunk 11]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
 	contract, ok := precompiles[common.Address(address)]
 	return contract, ok
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// State precompiled contract
+
+// DriverAddress is the NodeDriver contract address
+var DriverAddress = tosca.Address(common.HexToAddress("0xd100a01e00000000000000000000000000000000"))
```
Comment (facuMH) [inline go/processor/floria/precompiled.go:62]:
```suggestion
var driverAddress = tosca.Address(common.HexToAddress("0xd100a01e00000000000000000000000000000000"))

func DriverAddress() tosca.Address { return driverAddress }
```
if I recall correctly, this would be a safer way of exposing a public const, that is not really a const. 
This could be applied to other public "consts"

File: go/processor/floria/precompiled.go (hunk 2 above)
Comment (facuMH) [inline go/processor/floria/precompiled.go:104]:
```suggestion
func handleStatePrecompiledContract(state tosca.WorldState, sender, receiver tosca.Address, input []byte, gas tosca.Gas) (tosca.CallResult, bool) {
```

File: go/processor/floria/precompiled.go [hunk 12]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
+
+func handleStatePrecompiledContract(state tosca.WorldState, sender tosca.Address, address tosca.Address, input []byte, gas tosca.Gas) (tosca.CallResult, bool) {
+	if address != StateContractAddress {
+		return tosca.CallResult{}, false
+	}
+	gas, err := statePrecompiledContract(state, sender, input, gas)
+	return tosca.CallResult{
+		Success: err == nil,
+		Output:  nil,
+		GasLeft: tosca.Gas(gas),
```
Comment (facuMH) [inline go/processor/floria/precompiled.go:112]: unnecessary cast, `gas` is already of type `tosca.Gas`.
nit: maybe use `gasLeft` for the return value of `statePrecompiledContract`?

File: go/processor/floria/precompiled.go [hunk 13]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
+		return stateCopyCode(state, input, gas)
+	} else if bytes.Equal(selector, swapCodeMethodID) {
+		return stateSwapCode(state, input, gas)
+	} else if bytes.Equal(selector, setStorageMethodID) {
+		return stateSetStorage(state, input, gas)
+	} else if bytes.Equal(selector, incNonceMethodID) {
+		return stateIncNonce(state, sender, input, gas)
+	}
+
+	return 0, fmt.Errorf("invalid method ID")
```
Comment (facuMH) [inline go/processor/floria/precompiled.go:137]: this error could be a `constErr`

File: go/processor/floria/processor.go (hunk 5 above)
Comment (facuMH) [reply go/processor/floria/processor.go:31]: also if they are not used in `integration_tests` package, they can be private.

**Review by facuMH — CHANGES_REQUESTED** [review]
a few suggestions and questions, but looking good.

**Review by LuisPH3 — COMMENTED** [review]
The code seems correct, I would apply some quality-of-life changes.
I have been searching for the EIPs defining these operations with pretty bad luck, any hints there?

File: go/processor/floria/precompiled.go (hunk 4 above)
Comment (simonlechner) [reply go/processor/floria/precompiled.go:174]: I removed the checks and added an unit test to ensure it works as expected.

File: go/processor/floria/precompiled_test.go (hunk 8 above)
Comment (simonlechner) [reply go/processor/floria/precompiled_test.go:352]: I added a test for invalid input.

File: go/processor/floria/precompiled_test.go [hunk 14]
```diff
@@ -80,3 +81,347 @@ func TestPrecompiled_AddressesAreHandledCorrectly(t *testing.T) {
...
+			ErrExecutionReverted,
+		},
+		"insufficientMemoryGas": {
+			make([]byte, 64),
+			32000,
+			ErrOutOfGas,
+		},
+		"successful": {
+			append(append(make([]byte, 12), 0x01), make([]byte, 19+32)...),
+			32000 + codeSize*203,
```
[the remark this answers is not in the document]
Comment (simonlechner) [reply go/processor/floria/precompiled_test.go:161]: I changed it to the constant.

File: go/processor/floria/precompiled.go [hunk 15]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
+		code1 = []byte{}
+	}
+	cost0 := tosca.Gas(len(code0)) * (CreateDataGas + MemoryGas)
+	cost1 := tosca.Gas(len(code1)) * (CreateDataGas + MemoryGas)
+	cost = (cost0 + cost1) / 2 // 50% discount because trie size won't increase after pruning
+	if gas < cost {
+		return 0, ErrOutOfGas
+	}
+	gas -= cost
+	if account0 != account1 {
```
[the remark this answers is not in the document]
Comment (simonlechner) [reply go/processor/floria/precompiled.go:216]: Yes, it was defined like this by the original implementation.

File: go/processor/floria/precompiled.go [hunk 16]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
+	if code0 == nil {
+		code0 = []byte{}
+	}
+	code1 := state.GetCode(account1)
+	if code1 == nil {
+		code1 = []byte{}
+	}
+	cost0 := tosca.Gas(len(code0)) * (CreateDataGas + MemoryGas)
+	cost1 := tosca.Gas(len(code1)) * (CreateDataGas + MemoryGas)
+	cost = (cost0 + cost1) / 2 // 50% discount because trie size won't increase after pruning
```
[the remark this answers is not in the document]
Comment (simonlechner) [reply go/processor/floria/precompiled.go:211]: It was defined like this by the original implementation.

File: go/processor/floria/precompiled.go (hunk 11 above)
Comment (simonlechner) [reply go/processor/floria/precompiled.go:62]: They are no consts because of the common.HexToAddress function which can not be called during compile time. Also they are public because they are used in the integration test package.

File: go/processor/floria/run_context.go (hunk 9 above)
Comment (simonlechner) [reply go/processor/floria/run_context.go:83]: Thank you for the heads up!

File: go/processor/floria/precompiled.go (hunk 3 above)
Comment (simonlechner) [reply go/processor/floria/precompiled.go:116]: I merged both functions and renamed the resulting one to `handleStateContract`, I think this is clear now.

File: go/processor/floria/precompiled.go (hunk 13 above)
Comment (simonlechner) [reply go/processor/floria/precompiled.go:137]: I will move the `constErr` definition from the lfvm to tosca to reuse it in a follow up PR.

File: go/integration_test/processor/precompiled_test.go [hunk 17]
```diff
@@ -15,14 +15,16 @@ import (
...
+	"github.com/Fantom-foundation/Tosca/go/processor/floria"
 	"github.com/Fantom-foundation/Tosca/go/tosca"
 	"github.com/Fantom-foundation/Tosca/go/tosca/vm"
 )
 
-func TestProcessor_AllPreCompiledContractsAreAvailable(t *testing.T) {
+func TestProcessor_PreCompiledContractsCanBeProcessed(t *testing.T) {
 	tests := map[string]struct {
 		address tosca.Address
 		input   []byte
```
[the remark this answers is not in the document]
Comment (simonlechner) [reply go/integration_test/processor/precompiled_test.go:30]: I don't think this is required and the code is shorter (therefore more readable) without the casts.

File: go/processor/floria/precompiled.go (hunk 2 above)
Comment (simonlechner) [reply go/processor/floria/precompiled.go:104]: Since I adapted @LuisPH3 suggestion of multi line arguments, I think the type should stay part of the argument.

File: go/processor/floria/precompiled.go [hunk 18]
```diff
@@ -48,3 +54,215 @@ func getPrecompiledContract(address tosca.Address, revision tosca.Revision) (get
...
...e\":\"nextValidatorIDs\",\"type\":\"uint256[]\"}],\"name\":\"sealEpochValidators\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"internalType\":\"uint256[]\",\"name\":\"offlineTimes\",\"type\":\"uint256[]\"},{\"internalType\":\"uint256[]\",\"name\":\"offlineBlocks\",\"type\":\"uint256[]\"},{\"internalType\":\"uint256[]\",\"name\":\"uptimes\",\"type\":\"uint256[]\"},{\"internalType\":\"uint256[]\",\"name\":\"originatedTxsFee\",\"type\":\"uint256[]\"}],\"name\":\"sealEpoch\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"constant\":false,\"inputs\":[{\"internalType\":\"uint256[]\",\"name\":\"offlineTimes\",\"type\":\"uint256[]\"},{\"internalType\":\"uint256[]\",\"name\":\"offlineBlocks\",\"type\":\"uint256[]\"},{\"internalType\":\"uint256[]\",\"name\":\"uptimes\",\"type\":\"uint256[]\"},{\"internalType\":\"uint256[]\",\"name\":\"originatedTxsFee\",\"type\":\"uint256[]\"},{\"internalType\":\"uint256\",\"name\":\"usedGas\",\"type\":\"uint256\"}],\"name\":\"sealEpochV1\",\"outputs\":[],\"payable\":false,\"stateMutability\":\"nonpayable\",\"type\":\"function\"}]"
```
[the remark this answers is not in the document]
Comment (simonlechner) [reply go/processor/floria/precompiled.go:68]: Discussed offline.

File: go/processor/floria/precompiled_test.go [hunk 19]
```diff
@@ -80,3 +81,347 @@ func TestPrecompiled_AddressesAreHandledCorrectly(t *testing.T) {
...
+			32000,
+			ErrExecutionReverted,
+		},
+		"insufficientMemoryGas": {
+			make([]byte, 64),
+			32000,
+			ErrOutOfGas,
+		},
+		"successful": {
+			append(append(make([]byte, 12), 0x01), make([]byte, 19+32)...),
```
[the remark this answers is not in the document]
Comment (simonlechner) [reply go/processor/floria/precompiled_test.go:160]: I added address 0x00 and 0x00 as constants, I think this should make the input more readable.

File: go/processor/floria/precompiled_test.go [hunk 20]
```diff
@@ -80,3 +81,347 @@ func TestPrecompiled_AddressesAreHandledCorrectly(t *testing.T) {
...
+			state := tosca.NewMockWorldState(ctrl)
+			if test.err == nil {
+				state.EXPECT().SetBalance(tosca.Address{}, tosca.Value{})
+			}
+
+			gas, err := stateSetBalance(state, test.sender, test.input, test.gas)
+			if err != test.err {
+				t.Errorf("unexpected error, want %v, got %v", test.err, err)
+			}
+			if gas != 0 {
```
[the remark this answers is not in the document]
Comment (simonlechner) [reply go/processor/floria/precompiled_test.go:130]: Yes to ensure that the returned gas is 0.

File: go/processor/floria/state_contract.go [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+	createDataGas        = tosca.Gas(200)
+	memoryGas            = tosca.Gas(3)
+	sStoreSetGasEIP2200  = tosca.Gas(20000)
+)
+
+// DriverAddress is the NodeDriver contract address
+var DriverAddress = tosca.Address(common.HexToAddress("0xd100a01e00000000000000000000000000000000"))
+
+// StateContractAddress is the EvmWriter pre-compiled contract address
+var StateContractAddress = tosca.Address(common.HexToAddress("0xd100ec0000000000000000000000000000000000"))
```
Comment (HerbertJordan) [inline go/processor/floria/state_contract.go:37]: Could those be private? Since those are not constants, they could be updated accidentally.

**Review by HerbertJordan — APPROVED** [review]

File: go/processor/floria/precompiled.go (hunk 11 above)
Comment (facuMH) [reply go/processor/floria/precompiled.go]: it's fine that they are not consts, but maybe make them function ?

**Review by HerbertJordan — DISMISSED** [review]

File: go/processor/floria/state_contract_test.go [hunk 22]
```diff
@@ -0,0 +1,460 @@
...
+
+	remainingGas = tosca.Gas(10)
+)
+
+func TestStateContract_StateSetBalance(t *testing.T) {
+	tests := map[string]struct {
+		input  []byte
+		gas    tosca.Gas
+		sender tosca.Address
+		err    error
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:32]: nit: expectedError

File: go/processor/floria/state_contract_test.go [hunk 23]
```diff
@@ -0,0 +1,460 @@
...
+			state := tosca.NewMockWorldState(ctrl)
+
+			wantGas := tosca.Gas(0)
+			if test.err == nil {
+				wantGas = remainingGas
+				state.EXPECT().SetBalance(tosca.Address{}, tosca.Value{})
+			}
+
+			gas, err := executeStateSetBalance(state, test.sender, test.input, test.gas)
+			if err != test.err {
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:72]: nit: 
```go
if got, want := err, test.expectedError; !errors.Is(got, want){
     t.Errorf("unexpected error %v, wanted %v", got,  want)
}
```

File: go/processor/floria/state_contract_test.go [hunk 24]
```diff
@@ -0,0 +1,460 @@
...
+				t.Errorf("unexpected error, want %v, got %v", test.err, err)
+			}
+			if gas != wantGas {
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:75]: If error was expected (or err != nil) do not check the rest

File: go/processor/floria/state_contract_test.go [hunk 25]
```diff
@@ -0,0 +1,460 @@
...
+			if test.err == nil {
+				wantGas = remainingGas
+				state.EXPECT().SetCode(address0x01, tosca.Code(make([]byte, codeSize)))
+			}
+
+			gas, err := executeStateContractCopyCode(state, test.input, test.gas)
+			if err != test.err {
+				t.Errorf("unexpected error, want %v, got %v", test.err, err)
+			}
+			if gas != wantGas {
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:127]: If error was expected (or err !=  nil)  do not check the rest

File: go/processor/floria/state_contract_test.go [hunk 26]
```diff
@@ -0,0 +1,460 @@
...
+
+				state.EXPECT().SetCode(address0x00, tosca.Code(make([]byte, codeSize)))
+				state.EXPECT().SetCode(address0x01, tosca.Code(make([]byte, codeSize)))
+			}
+
+			gas, err := executeStateContractSwapCode(state, test.input, test.gas)
+			if err != test.err {
+				t.Errorf("unexpected error, want %v, got %v", test.err, err)
+			}
+			if gas != wantGas {
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:182]: If error was expected (or err != nil) do not check the rest

File: go/processor/floria/state_contract_test.go [hunk 27]
```diff
@@ -0,0 +1,460 @@
...
+			wantGas := tosca.Gas(0)
+			if test.err == nil {
+				wantGas = remainingGas
+				state.EXPECT().SetStorage(tosca.Address{}, tosca.Key{}, tosca.Word{})
+			}
+
+			gas, err := executeStateContractSetStorage(state, test.input, test.gas)
+			if err != test.err {
+				t.Errorf("unexpected error, want %v, got %v", test.err, err)
+			}
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:226]: If error was expected (or err != nil) do not check the rest

File: go/processor/floria/state_contract_test.go [hunk 28]
```diff
@@ -0,0 +1,460 @@
...
+				wantGas = remainingGas
+
+				state.EXPECT().GetNonce(tosca.Address{}).Return(uint64(5))
+				state.EXPECT().SetNonce(tosca.Address{}, uint64(5+increment))
+			}
+
+			gas, err := executeStateContractIncNonce(state, test.sender, test.input, test.gas)
+			if err != test.err {
+				t.Errorf("unexpected error, want %v, got %v", test.err, err)
+			}
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:290]: If error was expected (or err != nil) do not check the rest

File: go/processor/floria/state_contract_test.go [hunk 29]
```diff
@@ -0,0 +1,460 @@
...
+
+			_, err := test.operation(state, input, gas)
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+		})
+	}
+}
+
+func TestStateContract_InvalidCall(t *testing.T) {
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:424]:
```suggestion
func TestStateContract_InvalidCallReportsFailure(t *testing.T) {
```

File: go/processor/floria/state_contract_test.go [hunk 30]
```diff
@@ -0,0 +1,460 @@
...
+				t.Errorf("unexpected error, want %v, got %v", test.err, err)
+			}
+			if gas != wantGas {
+				t.Errorf("unexpected gas, want %v, got %v", wantGas, gas)
+			}
+		})
+	}
+}
+
+func TestStateContract_StateSwapCode(t *testing.T) {
```
Comment (LuisPH3) [inline go/processor/floria/state_contract_test.go:134]: I am afraid that the signature has changed: executeStateContractSwapCode

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
Please have a look at the test names.

File: go/processor/floria/state_contract_test.go (hunk 24 above)
Comment (simonlechner) [reply go/processor/floria/state_contract_test.go:75]: If an error is expected, it is important to return 0 gas that's why it is checked in all cases.

Comment (LuisPH3) [reply go/processor/floria/state_contract_test.go:76]: ok, right. When returning `err != nil` value should be zero initialized.

File: go/processor/floria/state_contract.go [hunk 31]
```diff
@@ -0,0 +1,246 @@
...
+	}
+}
+
+var ErrExecutionReverted = fmt.Errorf("execution reverted")
+var ErrOutOfGas = fmt.Errorf("out of gas")
+
+// handleStateContract is a reworked version of the original function from the Opera client.
+// It is used to handle epochs and allows to set balance, copy code, swap code, set storage, and increment nonce.
+// Source: https://github.com/Fantom-foundation/Sonic/blob/main/opera/contracts/evmwriter/evm_writer.go#L24
+func handleStateContract(
```
Comment (facuMH) [inline go/processor/floria/state_contract.go:85]: no need to change it now, but honestly, for function interface I think is better to have them in one line or multiline if over 80 columns. I agree that for test cases a liner per parameter is easier to read. But this to me seems clunky. it is just a personal opinion.

**Review by facuMH — APPROVED** [review]

File: go/processor/floria/state_contract.go (hunk 31 above)
Comment (LuisPH3) [reply go/processor/floria/state_contract.go:85]: It just don't fit in the side-by-side Github diff. And it turns quite difficult to read. That is it.

**Review by LuisPH3 — APPROVED** [review]
